# Disagreement Analysis: Human Annotators vs LLM Annotators

This notebook measures how well LLM annotators agree with human annotators when classifying Linux commits. Agreement is quantified with three complementary metrics:

| Metric | Full name | What it captures |
|--------|-----------|-----------------|
| **kA** | Krippendorff's Alpha | Ordinal agreement across all annotators simultaneously |
| **cK** | Cohen's Kappa (quadratic) | Pairwise ordinal agreement between two annotators |
| **aT** | Alt-Test score | Probability that the LLM is closer to human consensus than a held-out human peer |

For kA and cK, the notebook reports two values per LLM:

- **Mean** — the absolute agreement level achieved by that LLM configuration.
- **Diff** — Mean minus the human-only baseline (positive = above human level, negative = below).

For aT, four values are reported:

- **Mean** — avg ρ_f (advantage probability) across leave-one-out human perspectives.
- **Baseline (rho_h)** — human-only reference: the analogous probability when a held-out human replaces the LLM.
- **Diff** — Mean − Baseline (positive = LLM matches humans better than a human peer).
- **WR (Winning Rate)** — fraction of per-human hypothesis tests for which the null is rejected after Benjamini–Yekutieli correction, averaged across labels. A value ≥ 0.5 corresponds to `can_replace = True`.

---

## Context modes

Each LLM run is configured with a `context_mode` that controls what information is included in the prompt. The available modes are:

- **`message`** — Base mode. Only the commit message is sent as context.
- **`diff`** — Extends `message` with the commit diff, changed-file list, and line-count stats.
- **`single-label`** — Instructs the model to assign a dominant category and keep all others at 0 (conservative single-category policy).
- **`few-shot`** — Adds a set of human-annotated commits as calibration examples before the actual commit.

Modes are combined with `+` to activate several behaviors at once (e.g., `diff+single-label+few-shot`).

## Rounds

Each **round** is one fixed `context_mode` configuration applied across all LLMs:

| Round | `context_mode` |
|-------|---------------|
| `r1`  | `message` |
| `r2`  | `single-label` |
| `r3`  | `single-label+few-shot` |
| `r4`  | `diff+single-label` |
| `r5`  | `diff+single-label+few-shot` |

---

## Study design

For each LLM × round combination, the notebook proceeds as follows:

1. **Restrict to shared commits** — keep only commits annotated by all three humans (A, B, C) *and* by that LLM.
2. **Compute human baseline** — calculate kA, cK, and aT using only the three human annotators on that shared subset.
3. **Compute LLM metrics:**
   - **kA Mean** = average Krippendorff's Alpha across the three leave-one-out replacement permutations {(LLM,B,C), (A,LLM,C), (A,B,LLM)}.
   - **kA Diff** = kA Mean − kA(A,B,C).
   - **cK Mean** = average quadratic Cohen's Kappa of the three human–LLM pairs: cK(A,LLM), cK(B,LLM), cK(C,LLM).
   - **cK Diff** = cK Mean − mean of cK(A,B), cK(A,C), cK(B,C).
   - **aT Mean** = `avg_advantage_prob` (mean ρ_f) from `perform_alt_test(..., epsilon=0.15)`.
   - **aT Baseline** = mean ρ_h across the three leave-one-out human comparisons.
   - **aT Diff** = aT Mean − aT Baseline.
   - **aT WR** = fraction of the 3 per-human hypothesis tests (LLM vs A, LLM vs B, LLM vs C) for which the null hypothesis is rejected after Benjamini–Yekutieli correction, averaged across labels. Per-label WR ∈ {0, 0.33, 0.67, 1}; the reported WR is the mean across labels.
4. **Repeat per label** — all metrics are computed separately for each label (**bfc, bpc, prc, nfc**) and averaged into an overall score.

The result table updates automatically when new LLM files are placed under `data/llm-annotator-results/rX/`.


## Table of Contents

1. [Data discovery and loading](#1-data-discovery-and-loading)  
2. [Krippendorff's Alpha (kA)](#2-krippendorffs-alpha-ka)  
   - 2.1 [Worked Example: kA on Real BFC Annotations](#21-worked-example-ka-on-real-bfc-annotations)  
3. [Cohen's Kappa (cK)](#3-cohens-kappa-ck)  
   - 3.1 [Worked Example: cK on Real BFC Annotations](#31-worked-example-ck-on-real-bfc-annotations)  
4. [Alt-Test (aT)](#4-alt-test-at)  
   - 4.1 [Worked Example: aT on Real BFC Annotations](#41-worked-example-at-on-real-bfc-annotations)  
5. [Human Baseline Agreement](#5-human-baseline-agreement)  
6. [Intuition check with synthetic scenarios](#6-intuition-check-with-synthetic-scenarios)  
7. [Compute per-LLM metrics for all rounds](#7-compute-per-llm-metrics-for-all-rounds)  
8. [Main result table](#8-main-result-table-mean-and-diff-auto-updated-by-rounds-and-llm-files)  
   - 8.1 [Krippendorff's Alpha (kA)](#81-krippendorffs-alpha-ka)  
   - 8.2 [Cohen's Kappa (cK)](#82-cohens-kappa-ck)  
   - 8.3 [Alt-Test (aT)](#83-alt-test-at)  
   - 8.4 [Combined table (all metrics)](#84-combined-table-all-metrics)  
9. [Analysis with Merged BFC Label (BFC = max(BFC, BPC))](#9-analysis-with-merged-bfc-label-bfc--maxbfc-bpc)  
   - 9.1 [Krippendorff's Alpha (kA)](#91-krippendorffs-alpha-ka)  
   - 9.2 [Cohen's Kappa (cK)](#92-cohens-kappa-ck)  
   - 9.3 [Alt-Test (aT)](#93-alt-test-at)  
   - 9.4 [Combined table (all metrics)](#94-combined-table-all-metrics)  
10. [Side-by-side comparison: Original labels vs Merged BFC label](#10-side-by-side-comparison-original-labels-vs-merged-bfc-label)  
   - 10.1 [Original labels (sections 1–8)](#101-original-labels-sections-18)  
   - 10.2 [Merged BFC label (section 9)](#102-merged-bfc-label-section-9)  

---

In [59]:
from pathlib import Path
import re
import sys
import warnings
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.metrics import cohen_kappa_score
import krippendorff

# Suppress SciPy RuntimeWarning about precision loss in moment calculations
warnings.filterwarnings("ignore", category=RuntimeWarning, message=".*Precision loss.*")

sns.set_theme(style="whitegrid", context="talk")
pd.set_option("display.max_columns", 200)
pd.set_option("display.width", 200)

## 1. Data discovery and loading

In [60]:
ROOT = Path.cwd()
if not (ROOT / "data").exists():
    ROOT = ROOT.parent

DATA_DIR = ROOT / "data"
HUMAN_DIR = DATA_DIR / "human-annotator-results"
LLM_ROOT = DATA_DIR / "llm-annotator-results"

# Import Alt-Test utilities from analysis/alt_test/alt_test.py
ANALYSIS_DIR = ROOT / "analysis"
if str(ANALYSIS_DIR) not in sys.path:
    sys.path.append(str(ANALYSIS_DIR))

from alt_test.alt_test import compute_advantage_probabilities, perform_alt_test

ALT_TEST_EPSILON = 0.15

LABELS = ["bfc", "bpc", "prc", "nfc"]
HUMANS = ["A", "B", "C"]

def normalize_columns(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()
    out.columns = [c.strip().lower() for c in out.columns]
    return out


def standardize_annotations_df(df: pd.DataFrame, annotator_label: str | None = None) -> pd.DataFrame:
    out = normalize_columns(df)

    if "commit_hash" not in out.columns and "hash" in out.columns:
        out = out.rename(columns={"hash": "commit_hash"})

    expected = {"commit_hash", "annotator", *LABELS}
    missing = expected - set(out.columns)
    if missing:
        raise ValueError(f"Missing columns {missing}")

    out = out[["commit_hash", "annotator", *LABELS]].copy()
    out["commit_hash"] = out["commit_hash"].astype(str).str.strip()

    if annotator_label is not None:
        out["annotator"] = annotator_label

    return out


def load_human_data(human_dir: Path) -> dict[str, pd.DataFrame]:
    mapping = {}
    for annotator in HUMANS:
        path = human_dir / f"annotations_{annotator}.csv"
        raw = pd.read_csv(path)
        mapping[annotator] = standardize_annotations_df(raw, annotator_label=annotator)
    return mapping


def llm_name_from_path(path: Path) -> str:
    name = path.stem
    name = re.sub(r"^annotations_", "", name)
    name = re.sub(r"^ollama_", "", name)
    name = re.sub(r"^copilot_", "", name)
    return name


def discover_llm_files(llm_root: Path, filter: bool = False) -> dict[str, dict[str, Path]]:
    """
    Descubre archivos de anotaciones LLM organizados por rondas.
    
    Args:
        llm_root: Directorio raíz con las anotaciones LLM
        filter: Si True, filtra solo los LLMs en SELECTED_LLMS
    
    Returns:
        Diccionario de rondas -> modelo -> path
    """
    rounds = {}
    for round_dir in sorted([p for p in llm_root.iterdir() if p.is_dir() and p.name.startswith("r")], key=lambda p: int(re.sub(r"\D", "", p.name) or 0)):
        model_map = {}
        for csv_path in sorted(round_dir.glob("annotations_*.csv")):
            model_name = llm_name_from_path(csv_path)
            # Filtrar si está activado y el modelo no está en la lista
            if filter and model_name not in SELECTED_LLMS:
                continue
            model_map[model_name] = csv_path
        rounds[round_dir.name] = model_map
    return rounds


human_data = load_human_data(HUMAN_DIR)
# Lista de LLMs para filtrar (opcional)
SELECTED_LLMS = [
    "gpt-oss_120b_858",
]
llm_files_by_round = discover_llm_files(LLM_ROOT, filter=True)

print("Detected rounds:", ", ".join(llm_files_by_round.keys()))
for r, files in llm_files_by_round.items():
    print(f"  {r}: {len(files)} LLM files")


Detected rounds: r1, r2, r3, r4, r5
  r1: 1 LLM files
  r2: 1 LLM files
  r3: 1 LLM files
  r4: 1 LLM files
  r5: 1 LLM files


## 2. Krippendorff's Alpha (kA)

Krippendorff's Alpha measures reliability of agreement among **multiple raters** for ordinal data.

#### How it works with ordinal data

When `level_of_measurement='ordinal'` is used, the algorithm accounts for the magnitude of disagreement:

- **0 vs 4**: Maximum disagreement (opposite extremes)
- **3 vs 4**: Minimal disagreement (adjacent values)
- **4 vs 4**: Perfect agreement

The conceptual formula is:

$$\alpha = 1 - \frac{D_o}{D_e}$$

where $D_o$ is the observed disagreement and $D_e$ is the disagreement expected by chance.

In this notebook, **alpha and kA refer to the same metric** (Krippendorff's Alpha).

#### What we report

For each LLM, we report both:

- **kA Mean (LLM)**

$$kA_{\text{Mean}} = \overline{kA}_{replacement}$$

where $\overline{kA}_{replacement}$ is the mean kA over the three replacement permutations  
$\{(LLM,B,C),\;(A,LLM,C),\;(A,B,LLM)\}$.

- **kA Diff (LLM vs Humans)**

$$\Delta kA = kA_{\text{Mean}} - kA_{A,B,C}$$

where $kA_{A,B,C}$ is the human-only baseline.

In [61]:
def aligned_label_vectors(df_left: pd.DataFrame, df_right: pd.DataFrame, label: str):
    left = df_left[["commit_hash", label]].rename(columns={label: "left"})
    right = df_right[["commit_hash", label]].rename(columns={label: "right"})
    merged = left.merge(right, on="commit_hash", how="inner").dropna()
    merged["left"] = merged["left"].astype(int)
    merged["right"] = merged["right"].astype(int)
    return merged["left"].to_numpy(), merged["right"].to_numpy(), merged


def alpha_for_group(data_by_annotator: dict[str, pd.DataFrame], label: str) -> float:
    """Compute Krippendorff's Alpha (ordinal) for a group of annotators on a single label."""
    base = None
    for annotator, df in data_by_annotator.items():
        current = df[["commit_hash", label]].rename(columns={label: annotator})
        base = current if base is None else base.merge(current, on="commit_hash", how="inner")

    base = base.dropna()
    if base.empty:
        return np.nan

    matrix = []
    annotator_cols = [c for c in base.columns if c != "commit_hash"]
    for col in annotator_cols:
        matrix.append(base[col].astype(int).to_numpy())

    matrix = np.array(matrix)

    # Krippendorff alpha is undefined when all values collapse to a single category.
    # Return NaN so downstream means can ignore that label/model combination.
    if np.unique(matrix).size <= 1:
        return np.nan

    return float(krippendorff.alpha(reliability_data=matrix, level_of_measurement="ordinal"))

### 2.1 Worked Example: kA on Real BFC Annotations

This example computes Krippendorff's Alpha directly from the **real annotation values** for the **50 validation commits** in the **BFC** category.

- Each list corresponds to one human annotator (**A**, **B**, and **C**).
- The values are the original ordinal ratings used in this study (not synthetic data).

In [62]:
krippendorff.alpha(reliability_data=[
    # [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 3, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 3, 0], # Annotator A, bfc
    # [0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 4, 3, 4, 4, 0, 0, 4, 0, 4, 0, 0, 3, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0], # Annotator B, bfc
    # [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 4, 4, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0]  # Annotator C, bfc
    [0, 0, 4], # A
    [0, 4, 3], # B
    [0, 4, 4]  # C
], level_of_measurement="ordinal")

np.float64(0.4444444444444444)

## 3. Cohen's Kappa (cK)

Cohen's Kappa measures pairwise agreement between **two raters**.

#### How it works with ordinal data

We use `weights='quadratic'`, which penalises larger disagreements more heavily:

- **1-point difference** (e.g. 3 vs 4): small penalty
- **2-point difference** (e.g. 2 vs 4): medium penalty
- **4-point difference** (e.g. 0 vs 4): maximum penalty

#### What we report

For each LLM, we report both:

- **cK Mean (LLM)**

$$cK_{\text{Mean}} = \overline{cK}_{human\text{-}LLM}$$

where $\overline{cK}_{human\text{-}LLM} = \text{mean of } cK(A,LLM),\,cK(B,LLM),\,cK(C,LLM)$.

- **cK Diff (LLM vs Humans)**

$$\Delta cK = cK_{\text{Mean}} - \overline{cK}_{human\text{-}human}$$

where $\overline{cK}_{human\text{-}human} = \text{mean of } cK(A,B),\,cK(A,C),\,cK(B,C)$.

In [63]:
def safe_nanmean(values) -> float:
    """Return nanmean without emitting warnings when all values are NaN."""
    arr = np.asarray(values, dtype=float)
    if arr.size == 0 or np.all(np.isnan(arr)):
        return np.nan
    return float(np.nanmean(arr))


def cohen_for_pair(df_a: pd.DataFrame, df_b: pd.DataFrame, label: str) -> float:
    """Compute Cohen's Kappa (quadratic weights) for a pair of annotators on a single label."""
    y1, y2, _ = aligned_label_vectors(df_a, df_b, label)
    if len(y1) == 0:
        return np.nan

    # Kappa is undefined when both vectors collapse to a single class.
    if np.unique(np.concatenate([y1, y2])).size <= 1:
        return np.nan
    return float(cohen_kappa_score(y1, y2, weights="quadratic"))

### 3.1 Worked Example: cK on Real BFC Annotations

This example computes Cohen's Kappa directly from the **real annotation values** for the **50 validation commits** in the **BFC** category.

- Each list corresponds to one human annotator (**A**, **B**, and **C**).
- The values are the original ordinal ratings used in this study (not synthetic data).
- The cell computes pairwise quadratic-weighted kappa (**A-B**, **A-C**, **B-C**) and their mean for a transparent, reproducible check of the cK calculation.

In [64]:
# Cohen's Kappa example using the same real BFC annotations (50 commits)
bfc_A = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 3, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 3, 0]
bfc_B = [0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 4, 3, 4, 4, 0, 0, 4, 0, 4, 0, 0, 3, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0]
bfc_C = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 4, 4, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0]

ck_ab = cohen_kappa_score(bfc_A, bfc_B, weights="quadratic")
ck_ac = cohen_kappa_score(bfc_A, bfc_C, weights="quadratic")
ck_bc = cohen_kappa_score(bfc_B, bfc_C, weights="quadratic")
ck_mean = np.mean([ck_ab, ck_ac, ck_bc])

print(f"BFC (50 commits) - Cohen's Kappa A-B: {ck_ab:.4f}")
print(f"BFC (50 commits) - Cohen's Kappa A-C: {ck_ac:.4f}")
print(f"BFC (50 commits) - Cohen's Kappa B-C: {ck_bc:.4f}")
print(f"BFC (50 commits) - Mean pairwise Cohen's Kappa: {ck_mean:.4f}")

BFC (50 commits) - Cohen's Kappa A-B: 0.8658
BFC (50 commits) - Cohen's Kappa A-C: 0.8322
BFC (50 commits) - Cohen's Kappa B-C: 0.8721
BFC (50 commits) - Mean pairwise Cohen's Kappa: 0.8567


## 4. Alt-Test (aT)

The **Alt-Test** measures whether an LLM annotator's ratings are statistically close to any individual human annotator's ratings, using an advantage-probability framework.

#### How it works

For each commit and each human rater $h_i$, an **advantage probability** $\rho$ is computed:

$$\rho_f = P\left(|LLM - \hat{h}| < |h_i - \hat{h}|\right) + \frac{1}{2} P\left(|LLM - \hat{h}| = |h_i - \hat{h}|\right)$$

where $\hat{h}$ is the mean of the remaining human raters.  
The analogous quantity $\rho_h$ is computed replacing the LLM with a held-out human.

#### What we compute

For each LLM, we compute:

- **aT Mean (LLM)** = `avg_advantage_prob` (mean $\rho_f$) returned by `perform_alt_test(..., epsilon=0.15)`.
- **aT Baseline** = mean $\rho_h$ across the three leave-one-out human comparisons.
- **aT Diff** = aT Mean (LLM) − aT Baseline.
- **aT WR (Winning Rate)** = fraction of the 3 per-human hypothesis tests (LLM vs A, LLM vs B, LLM vs C) for which the null hypothesis is rejected after Benjamini–Yekutieli correction, averaged across labels. Per-label WR ∈ {0, 0.33, 0.67, 1}; the reported WR is the mean across labels. A value ≥ 0.5 corresponds to `can_replace = True` in the Alt-Test output.

A positive $\Delta aT$ means the LLM is *closer* to the human consensus than a human peer would be.


In [65]:
def intersect_commits_for_humans_and_llm(human_map: dict[str, pd.DataFrame], llm_df: pd.DataFrame) -> set[str]:
    """Return the set of commit hashes annotated by all humans AND the given LLM."""
    common = set(llm_df["commit_hash"].astype(str))
    for h in HUMANS:
        common &= set(human_map[h]["commit_hash"].astype(str))
    return common


def subset_by_commits(df: pd.DataFrame, commit_set: set[str]) -> pd.DataFrame:
    return df[df["commit_hash"].astype(str).isin(commit_set)].copy()


def build_alt_input_frame(
    hA: pd.DataFrame,
    hB: pd.DataFrame,
    hC: pd.DataFrame,
    lM: pd.DataFrame,
    label: str,
    llm_name: str,
    commit_set: set[str],
) -> pd.DataFrame:
    """Build wide dataframe required by compute_advantage_probabilities."""
    a_df = subset_by_commits(hA, commit_set)[["commit_hash", label]].rename(columns={label: "A"})
    b_df = subset_by_commits(hB, commit_set)[["commit_hash", label]].rename(columns={label: "B"})
    c_df = subset_by_commits(hC, commit_set)[["commit_hash", label]].rename(columns={label: "C"})
    llm_df = subset_by_commits(lM, commit_set)[["commit_hash", label]].rename(columns={label: llm_name})

    out = a_df.merge(b_df, on="commit_hash", how="inner")
    out = out.merge(c_df, on="commit_hash", how="inner")
    out = out.merge(llm_df, on="commit_hash", how="inner")
    out = out.dropna()

    # alt_test utility uses generic row iteration and does not require commit_hash semantics.
    return out.rename(columns={"commit_hash": "hash"})


def alt_test_metrics(
    hA: pd.DataFrame,
    hB: pd.DataFrame,
    hC: pd.DataFrame,
    lM: pd.DataFrame,
    label: str,
    llm_name: str,
    commit_set: set[str],
) -> tuple[float, float, float]:
    """Compute Alt-Test means for the LLM side (rho_f), human side (rho_h), and winning rate (WR)."""
    alt_input = build_alt_input_frame(hA, hB, hC, lM, label, llm_name, commit_set)
    if alt_input.empty:
        return np.nan, np.nan, np.nan

    advantage_results = compute_advantage_probabilities(
        alt_input,
        llm_name=llm_name,
        human_annotators=HUMANS,
    )
    alt_result = perform_alt_test(
        advantage_results,
        epsilon=ALT_TEST_EPSILON,
    )
    at_llm_mean = float(alt_result["avg_advantage_prob"])
    at_human_mean = float(np.mean([res["rho_h"] for res in alt_result["test_results"].values()]))
    at_wr = float(alt_result["winning_rate"])
    return at_llm_mean, at_human_mean, at_wr


### 4.1 Worked Example: aT on Real BFC Annotations

This example applies Alt-Test to the **real annotation values** for the **50 validation commits** in the **BFC** category.

- We reuse the same BFC vectors shown in the previous worked examples.
- Alt-Test requires three human annotators and one LLM annotator column.
- For demonstration, we use real results from `gpt-oss:20b`, so the full pipeline (`compute_advantage_probabilities` + `perform_alt_test`) can be inspected end-to-end.

In the output:

- **rho_f**: for each held-out human perspective, this is the probability that the **LLM** is closer to the human reference (the mean of the other humans), with ties counted as 0.5.
- **rho_h**: for the same perspective, this is the analogous probability when replacing the LLM with a **held-out human** (human-vs-human reference).

So, `avg_advantage_prob` is the mean of the `rho_f` values, and comparing `rho_f` with `rho_h` tells you whether the LLM behaves more like a human peer or deviates from human agreement.

In [66]:
# Alt-Test example using the same real BFC annotations (50 commits)
bfc_A = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 3, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 3, 0]
bfc_B = [0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 4, 3, 4, 4, 0, 0, 4, 0, 4, 0, 0, 3, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0]
bfc_C = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 4, 4, 4, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0]

# Demo LLM column from the same real BFC value space 
bfc_LLM = [4, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 3, 4, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 0, 4, 0, 4, 4, 4, 0, 0, 0, 0, 4, 4, 0, 1, 0, 0, 0]

alt_demo_df = pd.DataFrame({
    "hash": [f"bfc_{i:02d}" for i in range(1,51)],
    "A": bfc_A,
    "B": bfc_B,
    "C": bfc_C,
    "LLM": bfc_LLM,
})

advantage_results = compute_advantage_probabilities(
    alt_demo_df,
    llm_name="LLM",
    human_annotators=["A", "B", "C"],
)
alt_result = perform_alt_test(advantage_results, epsilon=ALT_TEST_EPSILON)

print(f"BFC (50 commits) - Alt-Test avg_advantage_prob (rho_f mean): {alt_result['avg_advantage_prob']:.4f}")
for annotator, values in alt_result["test_results"].items():
    print(f"  {annotator}: rho_f={values['rho_f']:.4f}, rho_h={values['rho_h']:.4f}")

BFC (50 commits) - Alt-Test avg_advantage_prob (rho_f mean): 0.6400
  A: rho_f=0.6400, rho_h=1.0000
  B: rho_f=0.6400, rho_h=0.9400
  C: rho_f=0.6400, rho_h=1.0000


## 5. Human Baseline Agreement

Before computing LLM metrics, we establish the **baseline** agreement among the three human annotators (A, B, C) over the validation set. This baseline serves as the reference for all **Diff** values in the main table.

| Baseline | Definition |
|----------|------------|
| **kA baseline** | $kA(A, B, C)$ — Krippendorff's Alpha among all three humans |
| **cK baseline** | mean of $cK(A,B),\;cK(A,C),\;cK(B,C)$ — average pairwise Cohen's Kappa |
| **aT baseline** | mean `avg_advantage_prob` across three leave-one-out human rotations (no LLM involved) |

The **aT baseline** is computed by rotating each human through the focal ("LLM") role and treating the other two as the human pool:

- Rotation 1: A as focal, pool = [B, C] → `avg_advantage_prob` (mean ρ_f)
- Rotation 2: B as focal, pool = [A, C] → `avg_advantage_prob`
- Rotation 3: C as focal, pool = [A, B] → `avg_advantage_prob`

The mean of these three values is the fixed human-only aT baseline, **independent of any LLM**.

A positive $\Delta$ (LLM Mean − baseline) means the LLM matches human agreement better than a human peer would; a negative $\Delta$ means it falls short.

In [67]:
def human_only_at_baseline(
    hA: pd.DataFrame,
    hB: pd.DataFrame,
    hC: pd.DataFrame,
    label: str,
    commit_set: set[str],
) -> float:
    """
    Compute a human-only aT baseline without any LLM.

    Each human takes the focal ('LLM') role in turn; the other two act as the human pool.
    avg_advantage_prob (mean rho_f) is collected for each of the three rotations and averaged.

    Rotations:
      A as focal, human pool = [B, C]
      B as focal, human pool = [A, C]
      C as focal, human pool = [A, B]
    """
    human_triplets = [
        ("A", hA, "B", hB, "C", hC),
        ("B", hB, "A", hA, "C", hC),
        ("C", hC, "A", hA, "B", hB),
    ]
    rho_f_values = []
    for focal_name, focal_df, h1_name, h1_df, h2_name, h2_df in human_triplets:
        f_sub  = subset_by_commits(focal_df, commit_set)[["commit_hash", label]].rename(columns={label: focal_name})
        h1_sub = subset_by_commits(h1_df,    commit_set)[["commit_hash", label]].rename(columns={label: h1_name})
        h2_sub = subset_by_commits(h2_df,    commit_set)[["commit_hash", label]].rename(columns={label: h2_name})

        wide = h1_sub.merge(h2_sub, on="commit_hash").merge(f_sub, on="commit_hash").dropna()
        wide = wide.rename(columns={"commit_hash": "hash"})

        if wide.empty or np.unique(wide[[h1_name, h2_name, focal_name]].values).size <= 1:
            continue

        adv = compute_advantage_probabilities(wide, llm_name=focal_name, human_annotators=[h1_name, h2_name])
        result = perform_alt_test(adv, epsilon=0.0)  # No epsilon needed since we're not comparing to an LLM; just want the raw advantage probabilities.
        rho_f_values.append(float(result["avg_advantage_prob"]))

    return float(np.mean(rho_f_values)) if rho_f_values else np.nan

In [68]:
# Use commits annotated by ALL humans (A, B, C) AND by AT LEAST ONE LLM.
# Note: Each LLM may have annotated a different subset of the 50 commits.
# kA/cK Diffs are computed per LLM against the LLM's own commit subset.
# The aT baseline is a fixed human-only value computed once on the full baseline_commit_set.

human_common_commits = (
    set(human_data["A"]["commit_hash"].astype(str))
    & set(human_data["B"]["commit_hash"].astype(str))
    & set(human_data["C"]["commit_hash"].astype(str))
)

llm_commit_sets = []
for model_files in llm_files_by_round.values():
    for csv_path in model_files.values():
        llm_tmp = standardize_annotations_df(pd.read_csv(csv_path))
        llm_commit_sets.append(set(llm_tmp["commit_hash"].astype(str)))

if not llm_commit_sets:
    raise ValueError("No LLM annotation files found to define the baseline commit set.")

# Check intersection vs union
llm_all_commits = set.intersection(*llm_commit_sets)
llm_any_commits = set.union(*llm_commit_sets)

print(f"LLM coverage analysis:")
print(f"  • Commits in ALL LLM files: {len(llm_all_commits)}")
print(f"  • Commits in ANY LLM file: {len(llm_any_commits)}")
if llm_all_commits != llm_any_commits:
    missing = len(llm_any_commits) - len(llm_all_commits)
    print(f"  • Commits in some but not all LLMs: {missing}")
print()

baseline_commit_set = human_common_commits & llm_any_commits

if not baseline_commit_set:
    raise ValueError("No shared commit set found between humans and LLM files for baseline.")

hA_base = subset_by_commits(human_data["A"], baseline_commit_set)
hB_base = subset_by_commits(human_data["B"], baseline_commit_set)
hC_base = subset_by_commits(human_data["C"], baseline_commit_set)

baseline_rows = []
for label in LABELS:
    ka_base      = alpha_for_group({"A": hA_base, "B": hB_base, "C": hC_base}, label)
    ck_ab        = cohen_for_pair(hA_base, hB_base, label)
    ck_ac        = cohen_for_pair(hA_base, hC_base, label)
    ck_bc        = cohen_for_pair(hB_base, hC_base, label)
    ck_base_mean = safe_nanmean([ck_ab, ck_ac, ck_bc])
    baseline_rows.append({
        "label": label,
        "kA (A,B,C)": round(ka_base, 3),
        "cK (A,B)": round(ck_ab, 3),
        "cK (A,C)": round(ck_ac, 3),
        "cK (B,C)": round(ck_bc, 3),
        "cK mean": round(ck_base_mean, 3),
    })

baseline_df = pd.DataFrame(baseline_rows).set_index("label")
baseline_df.loc["overall"] = baseline_df.mean().round(3)

# Human-only aT baseline: rotate each human through the focal ('LLM') role, no LLM involved.
print("Computing human-only aT baseline (3 rotations per label) ...")
at_human_baseline_by_label = {
    label: human_only_at_baseline(hA_base, hB_base, hC_base, label, baseline_commit_set)
    for label in LABELS
}
at_human_baseline_overall = safe_nanmean(list(at_human_baseline_by_label.values()))

baseline_df["aT (human)"] = pd.Series(
    {**{l: round(v, 3) for l, v in at_human_baseline_by_label.items()},
     "overall": round(at_human_baseline_overall, 3)}
)

print(f"\nHuman baseline (over {len(baseline_commit_set)} commits):")
print(f"Note: kA/cK Diff is per LLM vs its own commit subset. aT Diff uses the fixed human-only aT baseline.\n")
display(baseline_df)


LLM coverage analysis:
  • Commits in ALL LLM files: 908
  • Commits in ANY LLM file: 908



Computing human-only aT baseline (3 rotations per label) ...

Human baseline (over 908 commits):
Note: kA/cK Diff is per LLM vs its own commit subset. aT Diff uses the fixed human-only aT baseline.



,"kA (A,B,C)","cK (A,B)","cK (A,C)","cK (B,C)",cK mean,aT (human)
label,,,,,,
bfc,0.836,0.833,0.858,0.855,0.849,0.943
bpc,0.543,0.351,0.512,0.706,0.523,0.885
prc,0.807,0.718,0.795,0.881,0.798,0.914
nfc,0.830,0.807,0.872,0.912,0.864,0.957
overall,0.754,0.677,0.759,0.838,0.758,0.924


## 6. Intuition check with synthetic scenarios

The next cell provides a small **sanity-check demonstration** of how the three agreement metrics behave under controlled conditions.

It defines a helper function, `show_scenario(...)`, that takes four annotation vectors:

- **A, B, C** — three human annotators
- **LLM** — one model annotator

For each scenario, the function computes:

- **kA** — Krippendorff's Alpha for the three humans, and again after replacing one human at a time with the LLM
- **cK** — mean quadratic Cohen's Kappa for human–human pairs and human–LLM pairs
- **aT** — Alt-Test scores comparing the LLM to the human reference, alongside the corresponding human baseline

The printed output shows both the **absolute metric values** and the **difference** between LLM agreement and human-only agreement.

### What the scenarios illustrate

1. **Full agreement**  
    Humans and the LLM all give very similar ratings, so all metrics should be high.

2. **Humans agree, LLM diverges**  
    The human annotators remain consistent, but the LLM disagrees strongly, so LLM-based scores should drop.

3. **Moderate disagreement across everyone**  
    No annotator agrees perfectly, which shows how the metrics behave in a noisier and more realistic setting.

This cell is only for **interpretation and intuition**. It uses **synthetic examples**, not the real commit annotation files used in the main analysis.

In [69]:
# Helper: compute and display kA, cK, and aT for a scenario with synthetic data
def show_scenario(name, A, B, C, LLM):
    ids = [f"c{i:02d}" for i in range(len(A))]

    # ── kA ─────────────────────────────────────────────────────────────────
    # Suppress SciPy RuntimeWarning about precision loss during krippendorff.alpha calculations
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", category=RuntimeWarning, message=".*Precision loss.*")
        
        ka_humans = float(krippendorff.alpha(np.array([A, B, C]),     level_of_measurement="ordinal"))
        ka_r1     = float(krippendorff.alpha(np.array([LLM, B, C]),   level_of_measurement="ordinal"))
        ka_r2     = float(krippendorff.alpha(np.array([A, LLM, C]),   level_of_measurement="ordinal"))
        ka_r3     = float(krippendorff.alpha(np.array([A, B, LLM]),   level_of_measurement="ordinal"))
    
    ka_llm    = float(np.nanmean([ka_r1, ka_r2, ka_r3]))

    # ── cK ─────────────────────────────────────────────────────────────────
    ck_ab    = cohen_kappa_score(A, B,   weights="quadratic")
    ck_ac    = cohen_kappa_score(A, C,   weights="quadratic")
    ck_bc    = cohen_kappa_score(B, C,   weights="quadratic")
    ck_human = float(np.mean([ck_ab, ck_ac, ck_bc]))

    ck_al    = cohen_kappa_score(A,   LLM, weights="quadratic")
    ck_bl    = cohen_kappa_score(B,   LLM, weights="quadratic")
    ck_cl    = cohen_kappa_score(C,   LLM, weights="quadratic")
    ck_llm   = float(np.mean([ck_al, ck_bl, ck_cl]))

    # ── aT ─────────────────────────────────────────────────────────────────
    alt_df = pd.DataFrame({"hash": ids, "A": A, "B": B, "C": C, "LLM": LLM})
    adv    = compute_advantage_probabilities(alt_df, llm_name="LLM", human_annotators=["A", "B", "C"])
    at_res = perform_alt_test(adv, epsilon=ALT_TEST_EPSILON)
    at_llm   = float(at_res["avg_advantage_prob"])
    at_human = float(np.mean([v["rho_h"] for v in at_res["test_results"].values()]))
    at_wr    = float(at_res["winning_rate"])

    # ── Output ──────────────────────────────────────────────────────────────
    print(f"{'='*62}")
    print(f"  {name}")
    print(f"{'='*62}")
    print(f"    A:   {A}")
    print(f"    B:   {B}")
    print(f"    C:   {C}")
    print(f"  LLM:   {LLM}")
    print()
    print(f"  kA (humans): {ka_humans:+.3f}  |  kA (LLM mean): {ka_llm:+.3f}  |  ΔkA: {ka_llm - ka_humans:+.3f}")
    print(f"  cK (humans): {ck_human:+.3f}  |  cK (LLM mean): {ck_llm:+.3f}  |  ΔcK: {ck_llm - ck_human:+.3f}")
    print(f"  aT (rho_h):  {at_human:+.3f}  |  aT (rho_f):    {at_llm:+.3f}  |  ΔaT: {at_llm - at_human:+.3f}  |  WR: {at_wr:.2f}")
    print()


# ── Scenario 1: Full agreement ───────────────────────────
show_scenario(
    "Scenario 1 — Full agreement",
    A   = [0,0,4,4,0,3,0,4,0,0, 0,4,0,0,0,3,0,4,0,0, 0,0,4,0,0,0,3,0,4,0],
    B   = [0,0,4,4,0,4,0,4,0,0, 0,4,0,0,0,4,0,4,0,0, 0,0,4,0,0,0,4,0,4,0],
    C   = [0,0,3,4,0,4,0,4,0,0, 0,4,0,0,0,4,0,4,0,0, 0,0,4,0,0,0,4,0,4,0],
    LLM = [0,0,4,3,0,4,0,4,0,0, 0,4,0,0,0,4,0,4,0,0, 0,0,4,0,0,0,4,0,4,0],
)

# ── Scenario 2: Humans agree, LLM diverges ───────────────
show_scenario(
    "Scenario 2 — Humans agree, LLM diverges",
    A   = [0,0,4,4,0,3,0,4,0,0, 0,4,0,0,0,3,0,4,0,0, 0,0,4,0,0,0,3,0,4,0],
    B   = [0,0,4,4,0,4,0,4,0,0, 0,4,0,0,0,4,0,4,0,0, 0,0,4,0,0,0,4,0,4,0],
    C   = [0,0,3,4,0,4,0,4,0,0, 0,4,0,0,0,4,0,4,0,0, 0,0,4,0,0,0,4,0,4,0],
    LLM = [3,0,0,1,3,0,0,4,0,0, 4,0,0,0,3,1,0,3,0,0, 2,0,4,0,4,0,1,0,1,0],
)

# ── Scenario 3: Moderate disagreement across all annotators─────
show_scenario(
    "Scenario 3 — Moderate disagreement",
    A   = [0,1,4,2,0,3,1,4,0,2, 1,2,3,0,4,1,0,2,3,1, 2,0,4,3,1,2,3,0,1,4],
    B   = [2,0,3,4,1,4,0,2,3,0, 1,3,2,1,4,0,2,3,1,2, 0,1,4,2,0,3,4,1,0,3],
    C   = [1,0,4,3,2,2,0,3,1,1, 0,1,4,2,1,3,0,3,2,1, 2,1,3,4,2,1,2,3,1,2],
    LLM = [1,1,3,2,1,3,1,3,1,2, 1,2,3,1,3,1,1,2,3,1, 2,0,3,3,1,2,3,1,1,3],
)

  Scenario 1 — Full agreement
    A:   [0, 0, 4, 4, 0, 3, 0, 4, 0, 0, 0, 4, 0, 0, 0, 3, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 3, 0, 4, 0]
    B:   [0, 0, 4, 4, 0, 4, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0]
    C:   [0, 0, 3, 4, 0, 4, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0]
  LLM:   [0, 0, 4, 3, 0, 4, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0]

  kA (humans): +0.978  |  kA (LLM mean): +0.980  |  ΔkA: +0.002
  cK (humans): +0.944  |  cK (LLM mean): +0.952  |  ΔcK: +0.007
  aT (rho_h):  +0.956  |  aT (rho_f):    +0.967  |  ΔaT: +0.011  |  WR: 1.00

  Scenario 2 — Humans agree, LLM diverges
    A:   [0, 0, 4, 4, 0, 3, 0, 4, 0, 0, 0, 4, 0, 0, 0, 3, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 3, 0, 4, 0]
    B:   [0, 0, 4, 4, 0, 4, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0]
    C:   [0, 0, 3, 4, 0, 4, 0, 4, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 4, 0, 0, 0, 4, 0, 4, 0]
  LLM:   [3, 0, 0, 1,

## 7. Compute per-LLM metrics for all rounds

In [70]:
rows = []
long_rows = []

for round_name, model_files in llm_files_by_round.items():
    for model_name, csv_path in model_files.items():
        llm_df = standardize_annotations_df(pd.read_csv(csv_path), annotator_label=model_name)

        commit_set = intersect_commits_for_humans_and_llm(human_data, llm_df)
        if not commit_set:
            print(f"Skipping {round_name}/{model_name}: no overlapping commits with humans")
            continue

        hA = subset_by_commits(human_data["A"], commit_set)
        hB = subset_by_commits(human_data["B"], commit_set)
        hC = subset_by_commits(human_data["C"], commit_set)
        lM = subset_by_commits(llm_df, commit_set)

        label_ka_llm_mean = {}
        label_ka_diff = {}
        label_kc_llm_mean = {}
        label_kc_diff = {}
        label_at_llm_mean = {}
        label_at_human_mean = {}
        label_at_diff = {}
        label_at_wr = {}

        for label in LABELS:
            # Baseline human kA
            ka_humans = alpha_for_group({"A": hA, "B": hB, "C": hC}, label)

            # Replacement permutations for this LLM
            ka_replace_1 = alpha_for_group({"LLM": lM, "B": hB, "C": hC}, label)
            ka_replace_2 = alpha_for_group({"A": hA, "LLM": lM, "C": hC}, label)
            ka_replace_3 = alpha_for_group({"A": hA, "B": hB, "LLM": lM}, label)
            ka_llm_mix_mean = safe_nanmean([ka_replace_1, ka_replace_2, ka_replace_3])

            ka_diff = ka_llm_mix_mean - ka_humans

            # Baseline human pairwise kC
            kc_ab = cohen_for_pair(hA, hB, label)
            kc_ac = cohen_for_pair(hA, hC, label)
            kc_bc = cohen_for_pair(hB, hC, label)
            kc_humans_mean = safe_nanmean([kc_ab, kc_ac, kc_bc])

            # Human-LLM pairs kC
            kc_al = cohen_for_pair(hA, lM, label)
            kc_bl = cohen_for_pair(hB, lM, label)
            kc_cl = cohen_for_pair(hC, lM, label)
            kc_llm_mean = safe_nanmean([kc_al, kc_bl, kc_cl])

            kc_diff = kc_llm_mean - kc_humans_mean

            # Alt-Test: LLM mean (rho_f); Diff uses fixed human-only baseline
            at_llm_mean, at_human_mean, at_wr = alt_test_metrics(
                hA=hA,
                hB=hB,
                hC=hC,
                lM=lM,
                label=label,
                llm_name=model_name,
                commit_set=commit_set,
            )
            at_diff = at_llm_mean - at_human_mean

            label_ka_llm_mean[label] = ka_llm_mix_mean
            label_ka_diff[label] = ka_diff
            label_kc_llm_mean[label] = kc_llm_mean
            label_kc_diff[label] = kc_diff
            label_at_llm_mean[label] = at_llm_mean
            label_at_human_mean[label] = at_human_mean
            label_at_diff[label] = at_diff
            label_at_wr[label] = at_wr

            long_rows.append({
                "round": round_name,
                "llm": model_name,
                "label": label,
                "ka_llm_mean": ka_llm_mix_mean,
                "ka_diff": ka_diff,
                "kc_llm_mean": kc_llm_mean,
                "kc_diff": kc_diff,
                "at_llm_mean": at_llm_mean,
                "at_human_mean": at_human_mean,
                "at_diff": at_diff,
                "at_wr": at_wr,
                "n_commits": len(commit_set),
            })

        rows.append({
            "round": round_name,
            "llm": model_name,
            "n_commits": len(commit_set),
            "ka_llm_mean_overall": safe_nanmean([label_ka_llm_mean[x] for x in LABELS]),
            "ka_diff_overall": safe_nanmean([label_ka_diff[x] for x in LABELS]),
            "kc_llm_mean_overall": safe_nanmean([label_kc_llm_mean[x] for x in LABELS]),
            "kc_diff_overall": safe_nanmean([label_kc_diff[x] for x in LABELS]),
            "at_llm_mean_overall": safe_nanmean([label_at_llm_mean[x] for x in LABELS]),
            "at_human_mean_overall": safe_nanmean([label_at_human_mean[x] for x in LABELS]),
            "at_diff_overall": safe_nanmean([label_at_diff[x] for x in LABELS]),
            "at_wr_overall": safe_nanmean([label_at_wr[x] for x in LABELS]),
            **{f"ka_llm_mean_{k}": v for k, v in label_ka_llm_mean.items()},
            **{f"ka_diff_{k}": v for k, v in label_ka_diff.items()},
            **{f"kc_llm_mean_{k}": v for k, v in label_kc_llm_mean.items()},
            **{f"kc_diff_{k}": v for k, v in label_kc_diff.items()},
            **{f"at_llm_mean_{k}": v for k, v in label_at_llm_mean.items()},
            **{f"at_human_mean_{k}": v for k, v in label_at_human_mean.items()},
            **{f"at_diff_{k}": v for k, v in label_at_diff.items()},
            **{f"at_wr_{k}": v for k, v in label_at_wr.items()},
        })

results_df = pd.DataFrame(rows).sort_values(["round", "llm"]).reset_index(drop=True)
results_long_df = pd.DataFrame(long_rows).sort_values(["round", "llm", "label"]).reset_index(drop=True)

# print(f"Computed metrics for {results_df.shape[0]} LLM entries across {results_df['round'].nunique()} rounds")
# results_df.head()


## 8. Main result table (Mean and Diff, auto-updated by rounds and LLM files)

In [71]:
# Build the combined multi-index DataFrame for display
ROUND_CONTEXT_MODES = {"r1": "message", "r2": "single-label", "r3": "single-label+few-shot", "r4": "diff+single-label", "r5": "diff+single-label+few-shot"}

combined_rows = {}

# Add human baseline row
baseline_row = {}
for round_name in llm_files_by_round.keys():
    for metric, col_mean, col_diff, col_base in [
        ("kA", "kA (A,B,C)", None, None),
        ("cK", "cK mean", None, None),
        ("aT", "aT (human)", None, None),
    ]:
        if metric == "kA":
            baseline_row[(round_name, "kA", "Mean")] = baseline_df.loc["overall", "kA (A,B,C)"]
            baseline_row[(round_name, "kA", "Diff")] = 0.0
        elif metric == "cK":
            baseline_row[(round_name, "cK", "Mean")] = baseline_df.loc["overall", "cK mean"]
            baseline_row[(round_name, "cK", "Diff")] = 0.0
        elif metric == "aT":
            baseline_row[(round_name, "aT", "Mean")] = baseline_df.loc["overall", "aT (human)"]
            baseline_row[(round_name, "aT", "Baseline (rho_h)")] = baseline_df.loc["overall", "aT (human)"]
            baseline_row[(round_name, "aT", "Diff")] = 0.0
            baseline_row[(round_name, "aT", "WR")] = np.nan
combined_rows["Humans (baseline)"] = baseline_row

# Add LLM rows
all_llms = sorted(results_df["llm"].unique())
for llm in all_llms:
    llm_row = {}
    for round_name in llm_files_by_round.keys():
        mask = (results_df["round"] == round_name) & (results_df["llm"] == llm)
        sub = results_df[mask]
        if sub.empty:
            for col in [(round_name, "kA", "Mean"), (round_name, "kA", "Diff"),
                        (round_name, "cK", "Mean"), (round_name, "cK", "Diff"),
                        (round_name, "aT", "Mean"), (round_name, "aT", "Baseline (rho_h)"), (round_name, "aT", "Diff"), (round_name, "aT", "WR")]:
                llm_row[col] = np.nan
        else:
            r = sub.iloc[0]
            llm_row[(round_name, "kA", "Mean")] = r["ka_llm_mean_overall"]
            llm_row[(round_name, "kA", "Diff")] = r["ka_diff_overall"]
            llm_row[(round_name, "cK", "Mean")] = r["kc_llm_mean_overall"]
            llm_row[(round_name, "cK", "Diff")] = r["kc_diff_overall"]
            llm_row[(round_name, "aT", "Mean")] = r["at_llm_mean_overall"]
            llm_row[(round_name, "aT", "Baseline (rho_h)")] = r["at_human_mean_overall"]
            llm_row[(round_name, "aT", "Diff")] = r["at_diff_overall"]
            llm_row[(round_name, "aT", "WR")] = r["at_wr_overall"]
    combined_rows[llm] = llm_row

combined = pd.DataFrame(combined_rows).T
combined.columns = pd.MultiIndex.from_tuples(combined.columns)
combined = combined.astype(float)


def style_single_metric(df, metric):
    """Styler for a per-metric subset of the combined table."""
    styles = pd.DataFrame("", index=df.index, columns=df.columns)
    rounds = [c[0] for c in df.columns]
    unique_rounds = list(dict.fromkeys(rounds))
    palette = ["#f7fbff", "#fffaf0"]
    bg_map = {r: palette[i % len(palette)] for i, r in enumerate(unique_rounds)}
    for col in df.columns:
        styles.loc[:, col] = f"background-color: {bg_map.get(col[0], '')}"
    for i, col in enumerate(df.columns):
        nxt = df.columns[i + 1][0] if i + 1 < len(df.columns) else None
        if nxt != col[0]:
            styles.loc[:, col] += ";border-right: 3px solid #333"
    if "Humans (baseline)" in df.index:
        for col in df.columns:
            styles.loc["Humans (baseline)", col] += ";background-color: #d9ead3; font-weight: bold"
    for col in df.columns:
        if metric == "aT" and col[2] == "Baseline (rho_h)":
            continue
        col_series = df[col].drop(labels=["Humans (baseline)"], errors="ignore")
        if col_series.empty or col_series.isna().all():
            continue
        best_idx = col_series.idxmax()
        styles.loc[best_idx, col] += ";background-color: #fff2cc; font-weight: bold"
    mean_cols = [c for c in df.columns if c[2] == "Mean"]
    if mean_cols:
        best_value = None
        best_cells = []
        for row in df.index:
            if row == "Humans (baseline)":
                continue
            for col in mean_cols:
                val = df.loc[row, col]
                if pd.isna(val):
                    continue
                if best_value is None or val > best_value:
                    best_value = val
                    best_cells = [(row, col)]
                elif val == best_value:
                    best_cells.append((row, col))
        for row, col in best_cells:
            styles.loc[row, col] += ";background-color: #cfe2f3; font-weight: bold"
    styles = styles.map(lambda s: s.lstrip(";"))
    return styles


def style_combined(df):
    styles = pd.DataFrame("", index=df.index, columns=df.columns)
    rounds = [c[0] for c in df.columns]
    unique_rounds = []
    for r in rounds:
        if r not in unique_rounds:
            unique_rounds.append(r)
    palette = ["#f7fbff", "#fffaf0"]
    bg_map = {r: palette[i % len(palette)] for i, r in enumerate(unique_rounds)}
    for col in df.columns:
        styles.loc[:, col] = f"background-color: {bg_map.get(col[0], '')}"
    for i, col in enumerate(df.columns):
        nxt = df.columns[i + 1][0] if i + 1 < len(df.columns) else None
        if nxt != col[0]:
            styles.loc[:, col] += ";border-right: 3px solid #333"
    if "Humans (baseline)" in df.index:
        for col in df.columns:
            styles.loc["Humans (baseline)", col] += ";background-color: #d9ead3; font-weight: bold"
    for col in df.columns:
        if col[1] == "aT" and col[2] == "Baseline (rho_h)":
            continue
        col_series = df[col].drop(labels=["Humans (baseline)"], errors="ignore")
        if col_series.empty or col_series.isna().all():
            continue
        best_idx = col_series.idxmax()
        styles.loc[best_idx, col] += ";background-color: #fff2cc; font-weight: bold"
    for metric in ["kA", "cK", "aT"]:
        metric_cols = [c for c in df.columns if c[1] == metric and c[2] == "Mean"]
        if not metric_cols:
            continue
        best_value = None
        best_cells = []
        for row in df.index:
            if row == "Humans (baseline)":
                continue
            for col in metric_cols:
                val = df.loc[row, col]
                if pd.isna(val):
                    continue
                if best_value is None or val > best_value:
                    best_value = val
                    best_cells = [(row, col)]
                elif val == best_value:
                    best_cells.append((row, col))
        for row, col in best_cells:
            styles.loc[row, col] += ";background-color: #cfe2f3; font-weight: bold"
    styles = styles.map(lambda s: s.lstrip(";"))
    return styles


### 8.1 Krippendorff's Alpha (kA)

Columns: **Mean** (average kA across the three leave-one-out replacement permutations) and **Diff** (Mean − human-only baseline).

Legend: Green = Human baseline. Yellow = Best per column. Blue = Global best kA Mean across all LLMs and rounds.

In [72]:
ka_df = combined.loc[:, combined.columns.get_level_values(1) == "kA"]
ka_df.style.apply(style_single_metric, metric="kA", axis=None).format("{:.3f}", na_rep="—")

**Conclusions (kA).** No LLM reaches the human inter-annotator agreement level (kA = 0.703). The best-performing LLM (gpt-oss:20b) achieves a kA Mean of 0.635 in round 3, followed deepseek-r1_32b (0.629 in round 4) and gpt-oss:120b in round 1 (0.624).

### 8.2 Cohen's Kappa (cK)

Columns: **Mean** (average quadratic Cohen's Kappa across human–LLM pairs) and **Diff** (Mean − human-only baseline).

Legend: Green = Human baseline. Yellow = Best per column. Blue = Global best cK Mean across all LLMs and rounds.

In [73]:
ck_df = combined.loc[:, combined.columns.get_level_values(1) == "cK"]
ck_df.style.apply(style_single_metric, metric="cK", axis=None).format("{:.3f}", na_rep="—")

**Conclusions (cK).** No LLM surpasses the human baseline (cK = 0.721). The best-performing LLM (deepseek-r1:32b) achieves a cK Mean of 0.659 in round 4, followed by gpt-oss:120b in round 1 (0.603) and deepseek-r1:70b in round 2 (0.605).


### 8.3 Alt-Test (aT)

Columns: **Mean** (avg ρ_f across leave-one-out human perspectives), **Baseline (rho_h)** (human-only reference), **Diff** (Mean − Baseline), and **WR** (Winning Rate).

**WR — Winning Rate.** Fraction of the 3 per-human hypothesis tests (LLM vs A, LLM vs B, LLM vs C) for which the null hypothesis is rejected after Benjamini–Yekutieli correction, averaged across the 4 labels. Because there are exactly 3 human comparisons per label, the per-label WR can only take the values {0, 0.33, 0.67, 1}; the **WR** column shown here is the mean of those 4 per-label values and can therefore take any multiple of 1/12 in [0, 1]. A value ≥ 0.5 corresponds to `can_replace = True` in the Alt-Test output.

Legend: Green = Human baseline. Yellow = Best per column (excluding Baseline). Blue = Global best aT Mean across all LLMs and rounds.


In [74]:
at_df = combined.loc[:, combined.columns.get_level_values(1) == "aT"]
at_df.style.apply(style_single_metric, metric="aT", axis=None).format("{:.3f}", na_rep="—")

**Conclusions (aT).** No LLM exceeds the per-run human baseline (all Diffs are negative), but the top models come remarkably close. The best-performing LLMs are **gpt-5-mini** (r2) and **gpt-oss:120b** (r3) (0.938).

Regarding **Winning Rate (WR)**: **gpt-oss_120b** has the highest mean WR (0.917) in r3. This means that in 91.7% of the per-human comparisons across labels, the Alt-Test null hypothesis (LLM is not closer to human consensus than a human peer) is rejected after correction, suggesting that gpt-oss:120b's annotations are statistically indistinguishable from human agreement in those cases. 

Others LLMs with a WR ≥ 0.5:
- **gpt-5-mini** (r1) — WR = 0.500
- **gpt-5-mini** (r2 and r3) — WR = 0.667
- **gpt-oss_120b** (r2) — WR = 0.500
- **gpt-oss_120b** (r4) — WR = 0.750
- **deepseek-r1_32b** (r4) — WR = 0.500
- **gemma4:31b** (r4) — WR = 0.583

The following table is provided to help understand the WR value obtained by gpt-oss:120. Specifically, the value for r3 (0.917) corresponds to the mean of [1.0, 0.667, 1.0, 1.0]

In [75]:
for metric in ["at_wr"]:
    display_df = (
        results_long_df[results_long_df["llm"] == "gpt-oss_120b"]
        .pivot_table(index=["round", "llm"], columns="label", values=metric, aggfunc="first")
        .round(3)
        .sort_index()
    )
    metric_label = {
        "at_wr": "AT_WR",
    }[metric]
    display(display_df)

,label
round,llm


### 8.4 Combined table (all metrics)

All three metrics (kA, cK, aT) shown together, grouped by round. Each round block contains the kA, cK, and aT columns side by side.

Legend: Green = Human baseline (except aT Baseline). Yellow = Best per column (except aT Baseline). Blue = Global best per metric (kA/cK/aT Mean).

In [76]:
(
    combined.style
    .apply(style_combined, axis=None)
    .format("{:.3f}", na_rep="—")
)

---

## 9. Analysis with Merged BFC Label (BFC = max(BFC, BPC))

In this section we repeat the full kA / cK / aT analysis with a **merged BFC label**.

For every annotator (human or LLM) the fused BFC score is defined as:

$$\text{BFC}_{\text{merged}} = \max(\text{BFC}_{\text{original}},\; \text{BPC}_{\text{original}})$$

That is, if an annotator rated a commit as BFC = 3 and BPC = 1, the merged BFC value is **3**. The BPC column is dropped after merging, leaving three labels: **bfc** (merged), **prc**, and **nfc**.

Everything else — study design, metrics, styling — is identical to sections 5–8.

### Sub-sections

- [9.1 Krippendorff's Alpha (kA)](#91-krippendorffs-alpha-ka)
- [9.2 Cohen's Kappa (cK)](#92-cohens-kappa-ck)
- [9.3 Alt-Test (aT)](#93-alt-test-at)
- [9.4 Combined table (all metrics)](#94-combined-table-all-metrics)


In [77]:
# ── Section 10: Merged BFC label setup ───────────────────────────────────────

MERGED_LABELS = ["bfc", "prc", "nfc"]


def merge_bfc_bpc(df: pd.DataFrame) -> pd.DataFrame:
    """Return a copy of df where bfc = max(bfc, bpc) and the bpc column is dropped."""
    out = df.copy()
    out["bfc"] = np.maximum(out["bfc"].astype(int), out["bpc"].astype(int))
    return out.drop(columns=["bpc"])


# Apply merge to all human annotation DataFrames (loaded in section 1)
merged_human_data = {k: merge_bfc_bpc(v) for k, v in human_data.items()}

# Reuse the same baseline commit set from section 5
merged_hA_base = subset_by_commits(merged_human_data["A"], baseline_commit_set)
merged_hB_base = subset_by_commits(merged_human_data["B"], baseline_commit_set)
merged_hC_base = subset_by_commits(merged_human_data["C"], baseline_commit_set)

# Human baseline (kA, cK) for merged labels
merged_baseline_rows = []
for label in MERGED_LABELS:
    ka_base      = alpha_for_group({"A": merged_hA_base, "B": merged_hB_base, "C": merged_hC_base}, label)
    ck_ab        = cohen_for_pair(merged_hA_base, merged_hB_base, label)
    ck_ac        = cohen_for_pair(merged_hA_base, merged_hC_base, label)
    ck_bc        = cohen_for_pair(merged_hB_base, merged_hC_base, label)
    ck_base_mean = safe_nanmean([ck_ab, ck_ac, ck_bc])
    merged_baseline_rows.append({
        "label": label,
        "kA (A,B,C)": round(ka_base, 3),
        "cK (A,B)": round(ck_ab, 3),
        "cK (A,C)": round(ck_ac, 3),
        "cK (B,C)": round(ck_bc, 3),
        "cK mean": round(ck_base_mean, 3),
    })

merged_baseline_df = pd.DataFrame(merged_baseline_rows).set_index("label")
merged_baseline_df.loc["overall"] = merged_baseline_df.mean().round(3)

# Human-only aT baseline for merged labels
print("Computing human-only aT baseline for merged labels ...")
merged_at_human_baseline_by_label = {
    label: human_only_at_baseline(merged_hA_base, merged_hB_base, merged_hC_base, label, baseline_commit_set)
    for label in MERGED_LABELS
}
merged_at_human_baseline_overall = safe_nanmean(list(merged_at_human_baseline_by_label.values()))

merged_baseline_df["aT (human)"] = pd.Series(
    {**{l: round(v, 3) for l, v in merged_at_human_baseline_by_label.items()},
     "overall": round(merged_at_human_baseline_overall, 3)}
)

print(f"\nMerged-label human baseline (over {len(baseline_commit_set)} commits):")
display(merged_baseline_df)


Computing human-only aT baseline for merged labels ...

Merged-label human baseline (over 908 commits):


,"kA (A,B,C)","cK (A,B)","cK (A,C)","cK (B,C)",cK mean,aT (human)
label,,,,,,
bfc,0.743,0.626,0.756,0.832,0.738,0.892
prc,0.807,0.718,0.795,0.881,0.798,0.914
nfc,0.830,0.807,0.872,0.912,0.864,0.957
overall,0.793,0.717,0.808,0.875,0.800,0.921


In [78]:
# ── Section 10: Compute per-LLM metrics (merged labels) ──────────────────────

merged_rows = []
merged_long_rows = []

for round_name, model_files in llm_files_by_round.items():
    for model_name, csv_path in model_files.items():
        raw_llm_df = standardize_annotations_df(pd.read_csv(csv_path), annotator_label=model_name)
        merged_llm_df = merge_bfc_bpc(raw_llm_df)

        commit_set = intersect_commits_for_humans_and_llm(merged_human_data, merged_llm_df)
        if not commit_set:
            print(f"Skipping {round_name}/{model_name}: no overlapping commits with humans")
            continue

        hA = subset_by_commits(merged_human_data["A"], commit_set)
        hB = subset_by_commits(merged_human_data["B"], commit_set)
        hC = subset_by_commits(merged_human_data["C"], commit_set)
        lM = subset_by_commits(merged_llm_df, commit_set)

        label_ka_llm_mean = {}
        label_ka_diff = {}
        label_kc_llm_mean = {}
        label_kc_diff = {}
        label_at_llm_mean = {}
        label_at_human_mean = {}
        label_at_diff = {}
        label_at_wr = {}

        for label in MERGED_LABELS:
            # Baseline human kA
            ka_humans = alpha_for_group({"A": hA, "B": hB, "C": hC}, label)

            # Replacement permutations for this LLM
            ka_replace_1 = alpha_for_group({"LLM": lM, "B": hB, "C": hC}, label)
            ka_replace_2 = alpha_for_group({"A": hA, "LLM": lM, "C": hC}, label)
            ka_replace_3 = alpha_for_group({"A": hA, "B": hB, "LLM": lM}, label)
            ka_llm_mix_mean = safe_nanmean([ka_replace_1, ka_replace_2, ka_replace_3])

            ka_diff = ka_llm_mix_mean - ka_humans

            # Baseline human pairwise cK
            kc_ab = cohen_for_pair(hA, hB, label)
            kc_ac = cohen_for_pair(hA, hC, label)
            kc_bc = cohen_for_pair(hB, hC, label)
            kc_humans_mean = safe_nanmean([kc_ab, kc_ac, kc_bc])

            # Human-LLM pairs cK
            kc_al = cohen_for_pair(hA, lM, label)
            kc_bl = cohen_for_pair(hB, lM, label)
            kc_cl = cohen_for_pair(hC, lM, label)
            kc_llm_mean = safe_nanmean([kc_al, kc_bl, kc_cl])

            kc_diff = kc_llm_mean - kc_humans_mean

            # Alt-Test
            at_llm_mean, at_human_mean, at_wr = alt_test_metrics(
                hA=hA,
                hB=hB,
                hC=hC,
                lM=lM,
                label=label,
                llm_name=model_name,
                commit_set=commit_set,
            )
            at_diff = at_llm_mean - at_human_mean

            label_ka_llm_mean[label] = ka_llm_mix_mean
            label_ka_diff[label] = ka_diff
            label_kc_llm_mean[label] = kc_llm_mean
            label_kc_diff[label] = kc_diff
            label_at_llm_mean[label] = at_llm_mean
            label_at_human_mean[label] = at_human_mean
            label_at_diff[label] = at_diff
            label_at_wr[label] = at_wr

            merged_long_rows.append({
                "round": round_name,
                "llm": model_name,
                "label": label,
                "ka_llm_mean": ka_llm_mix_mean,
                "ka_diff": ka_diff,
                "kc_llm_mean": kc_llm_mean,
                "kc_diff": kc_diff,
                "at_llm_mean": at_llm_mean,
                "at_human_mean": at_human_mean,
                "at_diff": at_diff,
                "at_wr": at_wr,
                "n_commits": len(commit_set),
            })

        merged_rows.append({
            "round": round_name,
            "llm": model_name,
            "n_commits": len(commit_set),
            "ka_llm_mean_overall": safe_nanmean([label_ka_llm_mean[x] for x in MERGED_LABELS]),
            "ka_diff_overall": safe_nanmean([label_ka_diff[x] for x in MERGED_LABELS]),
            "kc_llm_mean_overall": safe_nanmean([label_kc_llm_mean[x] for x in MERGED_LABELS]),
            "kc_diff_overall": safe_nanmean([label_kc_diff[x] for x in MERGED_LABELS]),
            "at_llm_mean_overall": safe_nanmean([label_at_llm_mean[x] for x in MERGED_LABELS]),
            "at_human_mean_overall": safe_nanmean([label_at_human_mean[x] for x in MERGED_LABELS]),
            "at_diff_overall": safe_nanmean([label_at_diff[x] for x in MERGED_LABELS]),
            "at_wr_overall": safe_nanmean([label_at_wr[x] for x in MERGED_LABELS]),
            **{f"ka_llm_mean_{k}": v for k, v in label_ka_llm_mean.items()},
            **{f"ka_diff_{k}": v for k, v in label_ka_diff.items()},
            **{f"kc_llm_mean_{k}": v for k, v in label_kc_llm_mean.items()},
            **{f"kc_diff_{k}": v for k, v in label_kc_diff.items()},
            **{f"at_llm_mean_{k}": v for k, v in label_at_llm_mean.items()},
            **{f"at_human_mean_{k}": v for k, v in label_at_human_mean.items()},
            **{f"at_diff_{k}": v for k, v in label_at_diff.items()},
            **{f"at_wr_{k}": v for k, v in label_at_wr.items()},
        })

merged_results_df = pd.DataFrame(merged_rows).sort_values(["round", "llm"]).reset_index(drop=True)
merged_results_long_df = pd.DataFrame(merged_long_rows).sort_values(["round", "llm", "label"]).reset_index(drop=True)


In [79]:
# ── Section 10: Build combined_merged multi-index DataFrame ──────────────────

combined_merged_rows = {}

# Human baseline row
merged_baseline_row = {}
for round_name in llm_files_by_round.keys():
    merged_baseline_row[(round_name, "kA", "Mean")] = merged_baseline_df.loc["overall", "kA (A,B,C)"]
    merged_baseline_row[(round_name, "kA", "Diff")] = 0.0
    merged_baseline_row[(round_name, "cK", "Mean")] = merged_baseline_df.loc["overall", "cK mean"]
    merged_baseline_row[(round_name, "cK", "Diff")] = 0.0
    merged_baseline_row[(round_name, "aT", "Mean")] = merged_baseline_df.loc["overall", "aT (human)"]
    merged_baseline_row[(round_name, "aT", "Baseline (rho_h)")] = merged_baseline_df.loc["overall", "aT (human)"]
    merged_baseline_row[(round_name, "aT", "Diff")] = 0.0
    merged_baseline_row[(round_name, "aT", "WR")] = np.nan
combined_merged_rows["Humans (baseline)"] = merged_baseline_row

# LLM rows
all_merged_llms = sorted(merged_results_df["llm"].unique())
for llm in all_merged_llms:
    llm_row = {}
    for round_name in llm_files_by_round.keys():
        mask = (merged_results_df["round"] == round_name) & (merged_results_df["llm"] == llm)
        sub = merged_results_df[mask]
        if sub.empty:
            for col in [(round_name, "kA", "Mean"), (round_name, "kA", "Diff"),
                        (round_name, "cK", "Mean"), (round_name, "cK", "Diff"),
                        (round_name, "aT", "Mean"), (round_name, "aT", "Baseline (rho_h)"), (round_name, "aT", "Diff"), (round_name, "aT", "WR")]:
                llm_row[col] = np.nan
        else:
            r = sub.iloc[0]
            llm_row[(round_name, "kA", "Mean")] = r["ka_llm_mean_overall"]
            llm_row[(round_name, "kA", "Diff")] = r["ka_diff_overall"]
            llm_row[(round_name, "cK", "Mean")] = r["kc_llm_mean_overall"]
            llm_row[(round_name, "cK", "Diff")] = r["kc_diff_overall"]
            llm_row[(round_name, "aT", "Mean")] = r["at_llm_mean_overall"]
            llm_row[(round_name, "aT", "Baseline (rho_h)")] = r["at_human_mean_overall"]
            llm_row[(round_name, "aT", "Diff")] = r["at_diff_overall"]
            llm_row[(round_name, "aT", "WR")] = r["at_wr_overall"]
    combined_merged_rows[llm] = llm_row

combined_merged = pd.DataFrame(combined_merged_rows).T
combined_merged.columns = pd.MultiIndex.from_tuples(combined_merged.columns)
combined_merged = combined_merged.astype(float)


### 9.1 Krippendorff's Alpha (kA)

Columns: **Mean** (average kA across the three leave-one-out replacement permutations) and **Diff** (Mean − human-only baseline). BFC label is the merged max(BFC, BPC).

Legend: Green = Human baseline. Yellow = Best per column. Blue = Global best kA Mean across all LLMs and rounds.


In [80]:
ka_merged_df = combined_merged.loc[:, combined_merged.columns.get_level_values(1) == "kA"]
ka_merged_df.style.apply(style_single_metric, metric="kA", axis=None).format("{:.3f}", na_rep="—")


**Conclusions (kA).** No LLM reaches the human inter-annotator agreement level (kA = 0.739). The best-performing LLM (qwen3-coder_30b) achieves a kA Mean of 0.736 in round 5 (so close to human level), followed by gpt-oss_120b in round 3 (0.731) and qwen3-coder_30b (again) in round 4 (0.723).

### 9.2 Cohen's Kappa (cK)

Columns: **Mean** (average quadratic Cohen's Kappa across human–LLM pairs) and **Diff** (Mean − human-only baseline). BFC label is the merged max(BFC, BPC).

Legend: Green = Human baseline. Yellow = Best per column. Blue = Global best cK Mean across all LLMs and rounds.


In [81]:
ck_merged_df = combined_merged.loc[:, combined_merged.columns.get_level_values(1) == "cK"]
ck_merged_df.style.apply(style_single_metric, metric="cK", axis=None).format("{:.3f}", na_rep="—")


**Conclusions (cK).** At least 3 LLMs surpass or match the human baseline (cK = 0.751). The best-performing LLM (qwen3-coder_30b) achieves a cK Mean of 0.766 in round 5, followed by deepseek-r1_32b in round 4 (0.754) and gpt-oss_120b in round 3 (0.751).

### 9.3 Alt-Test (aT)

Columns: **Mean** (avg ρ_f across leave-one-out human perspectives), **Baseline (rho_h)** (human-only reference), **Diff** (Mean − Baseline), and **WR** (Winning Rate). BFC label is the merged max(BFC, BPC).

**WR — Winning Rate.** Fraction of the 3 per-human hypothesis tests (LLM vs A, LLM vs B, LLM vs C) for which the null hypothesis is rejected after Benjamini–Yekutieli correction, averaged across the 3 merged labels. Because there are exactly 3 human comparisons per label, the per-label WR can only take the values {0, 0.33, 0.67, 1}; the **WR** column shown here is the mean of those 3 per-label values and can therefore take any multiple of 1/9 in [0, 1]. A value ≥ 0.5 corresponds to `can_replace = True` in the Alt-Test output.

Legend: Green = Human baseline. Yellow = Best per column (excluding Baseline). Blue = Global best aT Mean across all LLMs and rounds.


In [82]:
at_merged_df = combined_merged.loc[:, combined_merged.columns.get_level_values(1) == "aT"]
at_merged_df.style.apply(style_single_metric, metric="aT", axis=None).format("{:.3f}", na_rep="—")

**Conclusions (aT).** No LLM exceeds the per-run human baseline (all Diffs are negative), but the top models come remarkably close. The best-performing LLM is **gpt-oss:120b** (r3) with aT = 0.949 (beeing the baseline in this case 0.951).

Regarding **Winning Rate (WR)**: **gpt-oss_120b** has the highest mean WR (0.889) in r3 and r4. This means that in 88.9% of the per-human comparisons across labels, the Alt-Test null hypothesis (LLM is not closer to human consensus than a human peer) is rejected after correction, suggesting that gpt-oss:120b's annotations are statistically indistinguishable from human agreement in those cases. 

Others LLMs with a WR ≥ 0.5:
- **gpt-oss_20b** (r3 and r4) — WR = 0.667
- **qwen3-coder_30b** (r3 and r5) — WR = 0.667
- **gpt-oss_120b** (r4) — WR = 0.667
- **gpt-5-mini** (r4) — WR = 0.556
- **deepseek-r1_32b** (r4) — WR = 0.500
- **deepseek-r1_70b** (r4) — WR = 0.583

### 9.4 Combined table (all metrics)

All three metrics (kA, cK, aT) shown together, grouped by round. BFC label is the merged max(BFC, BPC).

Legend: Green = Human baseline (except aT Baseline). Yellow = Best per column (except aT Baseline). Blue = Global best per metric (kA/cK/aT Mean).


In [83]:
(
    combined_merged.style
    .apply(style_combined, axis=None)
    .format("{:.3f}", na_rep="—")
)


---

## 10. Side-by-side comparison: Original labels vs Merged BFC label

### 10.1 Original labels (sections 1–8)

All three metrics for the four original labels (bfc, bpc, prc, nfc).

Legend: Green = Human baseline (except aT Baseline). Yellow = Best per column (except aT Baseline). Blue = Global best per metric (kA/cK/aT Mean).


In [84]:
(
    combined.style
    .apply(style_combined, axis=None)
    .format("{:.3f}", na_rep="—")
)


### 10.2 Merged BFC label (section 9)

All three metrics for the merged labels (bfc = max(bfc, bpc), prc, nfc).

Legend: Green = Human baseline (except aT Baseline). Yellow = Best per column (except aT Baseline). Blue = Global best per metric (kA/cK/aT Mean).


In [85]:
(
    combined_merged.style
    .apply(style_combined, axis=None)
    .format("{:.3f}", na_rep="—")
)
